### PR TITLE
feat(ccc): reuse a buffer for json encoding

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

...

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

**Review by colinlyguo — COMMENTED** [review]
LGTM. some benchmark of `ApplyTransaction` (maybe in a dev machine) would be helpful to show to optimization result.

File: rollup/circuitcapacitychecker/impl.go [hunk 1]
```diff
@@ -65,13 +67,14 @@ func (ccc *CircuitCapacityChecker) ApplyTransaction(traces *types.BlockTrace) (*
 		return nil, ErrUnknown
 	}
 
-	tracesByt, err := json.Marshal(traces)
+	ccc.jsonBuffer.Reset()
+	err := json.NewEncoder(&ccc.jsonBuffer).Encode(traces)
```
Comment (georgehao) [inline rollup/circuitcapacitychecker/impl.go:71]: does it cause race conditions between `ApplyTransaction` and `ApplyBlock` ?

Comment (omerfirmak) [reply rollup/circuitcapacitychecker/impl.go:71]: Nope, both functions require the ccc lock to be held.

**Comment by omerfirmak** [comment]
> LGTM. some benchmark of `ApplyTransaction` (maybe in a dev machine) would be helpful to show to optimization result.

Profiles show that GC churn decreases, but can't really quantify the improvement since there are a lot going on.

File: rollup/circuitcapacitychecker/impl.go (hunk 1 above)
Comment (georgehao) [reply rollup/circuitcapacitychecker/impl.go:71]: o, ic.

**Review by georgehao — DISMISSED** [review]

**Review by 0xmountaintop — DISMISSED** [review]

**Review by colinlyguo — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

File: rollup/circuitcapacitychecker/impl.go [hunk 2]
```diff
@@ -65,13 +67,14 @@ func (ccc *CircuitCapacityChecker) ApplyTransaction(traces *types.BlockTrace) (*
 		return nil, ErrUnknown
 	}
 
-	tracesByt, err := json.Marshal(traces)
+	ccc.jsonBuffer.Reset()
```
Comment (Thegaram) [inline rollup/circuitcapacitychecker/impl.go:70]: What if we encounter an exceptionally large trace, would we just keep the allocated buffer, even though we almost never fully utilize it?

Comment (omerfirmak) [reply rollup/circuitcapacitychecker/impl.go:70]: Yeah, that is the trade-off here. Such peaks of memory usage become sticky.

Comment (0xmountaintop) [reply rollup/circuitcapacitychecker/impl.go:70]: After disabling stack, I think we won't have such large traces as before.

**Pull request merged**